ow_faults/src/FaultInjector.cpp

```diff
@@ -75,16 +78,78 @@ void FaultInjector::setFaultsMessage(ow_faults::PTFaults& msg, ComponentFaults v
-    powerTemperatureOverloadValue = thermal_val;
-  } else if (!getTempBool) {
+void FaultInjector::powerFaultCb(){
+  std_msgs::Float64 soc_msg;
+  std_msgs::Float64 thermal_msg;
+  if (m_faults.low_state_of_charge_power_failure || m_faults.instantaneous_capacity_loss_power_failure) {
+    if (m_faults.low_state_of_charge_power_failure && m_faults.instantaneous_capacity_loss_power_failure) {
+      //both faults are on
+      setPowerFaultValues("SOC", 0.05, .06);    
+    } 
+    else if (m_faults.low_state_of_charge_power_failure){
+      // Fault is a range ( anything < 10%)
+      setPowerFaultValues("SOC-LOW", 0.001, .01);
+    } 
+    else if (m_faults.instantaneous_capacity_loss_power_failure) {
+      // (most recent and current). If the % difference is > 5% and no other tasks in progress, then fault. 
+      setPowerFaultValues("SOC-CAP-LOSS", 0.05, .06);
+    } 
+    soc_msg.data = powerStateOfChargeValue;
+    m_fault_power_state_of_charge_pub.publish(soc_msg);
+  } 
+  else { //no faults, set to SOC
+      powerStateOfChargeValue = originalSOC;
+      soc_msg.data = powerStateOfChargeValue;
+      m_fault_power_state_of_charge_pub.publish(soc_msg);
+  }
+  if(m_faults.thermal_power_failure){
+    // if > 50 degrees C, then consider fault. 
+    setPowerFaultValues("THERMAL", 50, 90);
+    thermal_msg.data = powerTemperatureOverloadValue;
+    m_fault_power_temp_pub.publish(thermal_msg);
+  } else {
     powerTemperatureOverloadValue = NAN;
   }
 }
 
+float FaultInjector::getRandomFloatFromRange( float min_val, float max_val){
+  return min_val + (max_val - min_val) * (rand() / float(RAND_MAX));
+}
+
+void FaultInjector::setPowerFaultValues(const string& powerType, float min_val, float max_val){
+  float percent_difference = getRandomFloatFromRange(powerStateOfChargeValue*min_val, powerStateOfChargeValue*max_val);
+  float min_power_default = 0.5;
+  float soc_fail_default = 10;
+  float newPowerStateOfChargeValue = ((powerStateOfChargeValue-percent_difference) < min_power_default) ?  min_power_default : (powerStateOfChargeValue-percent_difference);
+
+  if (powerType == "SOC" || powerType == "SOC-LOW"){
+    // value needs to be between 0 and 9.99, decreased until 0
+    if (powerStateOfChargeValue > soc_fail_default){
+      powerStateOfChargeValue = soc_fail_default;
+    } else {
+      powerStateOfChargeValue = newPowerStateOfChargeValue;
+    }
+  }
+  else if (powerType == "SOC-CAP-LOSS"){
+    // cap loss error have increase of 5-10 % of the SOC
+    powerStateOfChargeValue = newPowerStateOfChargeValue;
+  }
+  else if (powerType == "THERMAL"){
+    if (isnan(powerTemperatureOverloadValue)){
+      powerTemperatureOverloadValue = getRandomFloatFromRange(min_val, max_val);
+    } else {
+      float change_in_val = getRandomFloatFromRange(.01, 1);
+      powerTemperatureOverloadValue += change_in_val;
+    }
+  }
+}
+
+void FaultInjector::powerSOCListener(const std_msgs::Float64& msg)
+{
+  originalSOC = msg.data;
+
+}
```

Comment:
Use the original topic callback as a trigger instead of a timer, this way you don't need to worry about what is the publish rate of the original topic currently is 
```suggestion
void FaultInjector::powerSOCListener(const std_msgs::Float64& msg)
{
  originalSOC = msg.data;
  powerFaultCb();
}
```